<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Gumbel &mdash; TRANSIT v3.2.4 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/default.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Griffin" href="method_griffin.html" />
    <link rel="prev" title="Quality Control" href="transit_features.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> TRANSIT
            <img src="_static/transit_logo.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                v3.2.4
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">TRANSIT MANUAL</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit_overview.html">TRANSIT Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_running.html">Running TRANSIT</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_features.html">Quality Control</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">ANALYSIS METHODS</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Gumbel</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#how-does-it-work">How does it work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#outputs-and-diagnostics">Outputs and diagnostics</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-time">Run-time</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="method_griffin.html">Griffin</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_tn5gaps.html">Tn5Gaps</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_HMM.html">HMM</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_resampling.html">Resampling</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_Utest.html">Mann-Whitney U-test (utest)</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_GI.html">Genetic Interactions</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_anova.html">ANOVA</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_zinb.html">ZINB</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_normalization.html">Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_pathway_enrichment.html">Pathway Enrichment Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_corrplot.html">corrplot</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_heatmap.html">heatmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_ttnfitness.html">TTN-Fitness</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">TRANSIT Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit_interactions_tutorial.html">Tutorial: Genetic Interactions Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_normalization_tutorial.html">Tutorial: Normalize datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_export_tutorial.html">Tutorial: Export datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_console_cheatsheet.html">Console Mode Cheat-Sheet</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">TPP Manual</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tpp.html">TPP Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#running-tpp">Running TPP</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#mapping-to-genomes-with-multiple-contigs">Mapping to Genomes with Multiple Contigs</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#overview-of-data-processing-procedure">Overview of Data Processing Procedure</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#statistics">Statistics</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Code Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit.html">transit package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">TRANSIT</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Gumbel</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/method_gumbel.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gumbel">
<span id="id1"></span><h1>Gumbel<a class="headerlink" href="#gumbel" title="Permalink to this headline"></a></h1>
<p>The Gumbel can be used to determine which genes are essential in a
single condition. It does a gene-by-gene analysis of the insertions at
TA sites with each gene, makes a call based on the longest consecutive
sequence of TA sites without insertion in the genes, calculates the
probability of this using a Bayesian model.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Intended only for <strong>Himar1</strong> datasets.</p>
</div>
<section id="how-does-it-work">
<h2>How does it work?<a class="headerlink" href="#how-does-it-work" title="Permalink to this headline"></a></h2>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p>This method for identifying essential genes is based on analyzing
‘gaps’, or consecutive sequences of TA sites lacking insertions.
The statistical significance of the length of a gap is determined
using the Gumbel distribution, which is a form of an Extreme-Value distribution.</p>
<p>For a formal description of how this method works, see our paper <a class="reference internal" href="transit_overview.html#dejesus2013" id="id2"><span>[DeJesus2013]</span></a>:</p>
<div class="line-block">
<div class="line">DeJesus, M.A., Zhang, Y.J., Sassettti, C.M., Rubin, E.J.,
Sacchettini, J.C., and Ioerger, T.R. (2013).  <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23361328">Bayesian analysis of
gene essentiality based on sequencing of transposon insertion
libraries.</a>
<em>Bioinformatics</em>, 29(6):695-703.</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p><strong>Update (2021) - Binomial</strong></p>
<p>Since the Gumbel method depends on the overall
saturation (percent of TA sites with insertions), it can sometimes
call a lot of smaller genes ‘Uncertain’.  This might reduce the total
number of essentials detected.  For example, if the saturation is
~30%, no genes with fewer than ~10 TA sites might confidently be
labeled as Essential.  In particular, there are often genes with no
insertions that look like they should obviously be called essential,
and yet, they are too short to be confidently called essential in
low-saturation datasets by the conservative Gumbel model.</p>
<p>To compensate for this, we have added a simple <strong>Binomial</strong> model for
detecting small genes totally lacking insertions (described in <a class="reference external" href="https://journals.asm.org/doi/full/10.1128/mSystems.00876-21">(Choudhery et al, 2021)</a>).  In
the output file, they are labled as ‘EB’ to distinguish them from
essentials (E) based on the Gumbel model.  The EB genes supplement
genes marked E by Gumbel, and the combination of the 2 groups (E and
EB) should be considered as ‘essentials’.  The number of genes in
each category is reported in a table in the header of the output file, like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="k">for</span> <span class="n">an</span> <span class="n">M</span><span class="o">.</span> <span class="n">tuberculosis</span> <span class="n">TnSeq</span> <span class="n">dataset</span> <span class="k">with</span> <span class="mi">60</span><span class="o">%</span> <span class="n">saturation</span><span class="p">:)</span>
<span class="c1">#Summary of Essentiality Calls:</span>
<span class="c1">#  E  =  551 (essential based on Gumbel)</span>
<span class="c1">#  EB =   94 (essential based on Binomial)</span>
<span class="c1">#  NE = 2829 (non-essential)</span>
<span class="c1">#  U  =  262 (uncertain)</span>
<span class="c1">#  S  =  254 (too short)</span>

<span class="p">(</span><span class="k">for</span> <span class="n">an</span> <span class="n">M</span><span class="o">.</span> <span class="n">tuberculosis</span> <span class="n">TnSeq</span> <span class="n">dataset</span> <span class="k">with</span> <span class="mi">40</span><span class="o">%</span> <span class="n">saturation</span><span class="p">:)</span>
<span class="c1">#  E  =  194 (essential based on Gumbel)</span>
<span class="c1">#  EB =  315 (essential based on Binomial)</span>
<span class="c1">#  NE = 2441 (non-essential)</span>
<span class="c1">#  U  =  774 (uncertain)</span>
<span class="c1">#  S  =  266 (too short)</span>
</pre></div>
</div>
</section>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">python3</span> <span class="n">transit</span><span class="o">.</span><span class="n">py</span> <span class="n">gumbel</span> <span class="o">&lt;</span><span class="n">comma</span><span class="o">-</span><span class="n">separated</span> <span class="o">.</span><span class="n">wig</span> <span class="n">files</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">annotation</span> <span class="o">.</span><span class="n">prot_table</span> <span class="ow">or</span> <span class="n">GFF3</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">output</span> <span class="n">file</span><span class="o">&gt;</span> <span class="p">[</span><span class="n">Optional</span> <span class="n">Arguments</span><span class="p">]</span>
      <span class="n">Optional</span> <span class="n">Arguments</span><span class="p">:</span>
      <span class="o">-</span><span class="n">s</span> <span class="o">&lt;</span><span class="n">integer</span><span class="o">&gt;</span>    <span class="o">:=</span>  <span class="n">Number</span> <span class="n">of</span> <span class="n">samples</span><span class="o">.</span> <span class="n">Default</span><span class="p">:</span> <span class="o">-</span><span class="n">s</span> <span class="mi">10000</span>
      <span class="o">-</span><span class="n">b</span> <span class="o">&lt;</span><span class="n">integer</span><span class="o">&gt;</span>    <span class="o">:=</span>  <span class="n">Number</span> <span class="n">of</span> <span class="n">Burn</span><span class="o">-</span><span class="ow">in</span> <span class="n">samples</span><span class="o">.</span> <span class="n">Default</span> <span class="o">-</span><span class="n">b</span> <span class="mi">500</span>
      <span class="o">-</span><span class="n">m</span> <span class="o">&lt;</span><span class="n">integer</span><span class="o">&gt;</span>    <span class="o">:=</span>  <span class="n">Smallest</span> <span class="n">read</span><span class="o">-</span><span class="n">count</span> <span class="n">to</span> <span class="n">consider</span><span class="o">.</span> <span class="n">Default</span><span class="p">:</span> <span class="o">-</span><span class="n">m</span> <span class="mi">1</span>
      <span class="o">-</span><span class="n">t</span> <span class="o">&lt;</span><span class="n">integer</span><span class="o">&gt;</span>    <span class="o">:=</span>  <span class="n">Trims</span> <span class="nb">all</span> <span class="n">but</span> <span class="n">every</span> <span class="n">t</span><span class="o">-</span><span class="n">th</span> <span class="n">value</span><span class="o">.</span> <span class="n">Default</span><span class="p">:</span> <span class="o">-</span><span class="n">t</span> <span class="mi">1</span>
      <span class="o">-</span><span class="n">r</span> <span class="o">&lt;</span><span class="n">string</span><span class="o">&gt;</span>     <span class="o">:=</span>  <span class="n">How</span> <span class="n">to</span> <span class="n">handle</span> <span class="n">replicates</span><span class="o">.</span> <span class="n">Sum</span> <span class="ow">or</span> <span class="n">Mean</span><span class="o">.</span> <span class="n">Default</span><span class="p">:</span> <span class="o">-</span><span class="n">r</span> <span class="n">Sum</span>
      <span class="o">-</span><span class="n">iN</span> <span class="o">&lt;</span><span class="nb">float</span><span class="o">&gt;</span>     <span class="o">:=</span>  <span class="n">Ignore</span> <span class="n">TAs</span> <span class="n">occuring</span> <span class="n">at</span> <span class="n">given</span> <span class="n">percentage</span> <span class="p">(</span><span class="k">as</span> <span class="n">integer</span><span class="p">)</span> <span class="n">of</span> <span class="n">the</span> <span class="n">N</span> <span class="n">terminus</span><span class="o">.</span> <span class="n">Default</span><span class="p">:</span> <span class="o">-</span><span class="n">iN</span> <span class="mi">0</span>
      <span class="o">-</span><span class="n">iC</span> <span class="o">&lt;</span><span class="nb">float</span><span class="o">&gt;</span>     <span class="o">:=</span>  <span class="n">Ignore</span> <span class="n">TAs</span> <span class="n">occuring</span> <span class="n">at</span> <span class="n">given</span> <span class="n">percentage</span> <span class="p">(</span><span class="k">as</span> <span class="n">integer</span><span class="p">)</span> <span class="n">of</span> <span class="n">the</span> <span class="n">C</span> <span class="n">terminus</span><span class="o">.</span> <span class="n">Default</span><span class="p">:</span> <span class="o">-</span><span class="n">iC</span> <span class="mi">0</span>
</pre></div>
</div>
</section>
<section id="parameters">
<h2>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p><strong>Samples:</strong> Gumbel uses Metropolis-Hastings (MH) to generate samples
of posterior distributions. The default setting is to run the
simulation for 10,000 iterations. This is usually enough to assure
convergence of the sampler and to provide accurate estimates of
posterior probabilities. Less iterations may work, but at the risk of
lower accuracy.</p></li>
<li><p><strong>Burn-In:</strong> Because the MH sampler many not have stabilized in the
first few iterations, a “burn-in” period is defined. Samples obtained
in this “burn-in” period are discarded, and do not count towards
estimates.</p></li>
<li><p><strong>Trim:</strong> The MH sampler produces Markov samples that are correlated.
This parameter dictates how many samples must be attempted for every
sampled obtained. Increasing this parameter will decrease the
auto-correlation, at the cost of dramatically increasing the
run-time. For most situations, this parameter should be left at the
default of “1”.</p></li>
<li><p><strong>Minimum Read:</strong> The minimum read count that is considered a true
read. Because the Gumbel method depends on determining gaps of TA
sites lacking insertions, it may be susceptible to spurious reads
(e.g. errors). The default value of 1 will consider all reads as true
reads. A value of 2, for example, will ignore read counts of 1.</p></li>
<li><p><strong>Replicates:</strong> Determines how to deal with replicates by averaging
the read-counts or summing read counts across datasets. This should
not have an affect for the Gumbel method, aside from potentially
affecting spurious reads.</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="outputs-and-diagnostics">
<h2>Outputs and diagnostics<a class="headerlink" href="#outputs-and-diagnostics" title="Permalink to this headline"></a></h2>
<p>The Gumbel method generates a tab-separated output file at the location
chosen by the user. This file will automatically be loaded into the
Results Files section of the GUI, allowing you to display it as a table.
Alternatively, the file can be opened in a spreadsheet software like
Excel as a tab-separated file. The columns of the output file are
defined as follows:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 12%" />
<col style="width: 88%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column Header</p></th>
<th class="head"><p>Column Definition</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ORF</p></td>
<td><p>Gene ID.</p></td>
</tr>
<tr class="row-odd"><td><p>Name</p></td>
<td><p>Name of the gene.</p></td>
</tr>
<tr class="row-even"><td><p>Description</p></td>
<td><p>Gene description.</p></td>
</tr>
<tr class="row-odd"><td><p>k</p></td>
<td><p>Number of Transposon Insertions Observed within the ORF.</p></td>
</tr>
<tr class="row-even"><td><p>n</p></td>
<td><p>Total Number of TA dinucleotides within the ORF.</p></td>
</tr>
<tr class="row-odd"><td><p>r</p></td>
<td><p>Length of the Maximum Run of Non-Insertions observed.</p></td>
</tr>
<tr class="row-even"><td><p>s</p></td>
<td><p>Span of nucleotides for the Maximum Run of Non-Insertions.</p></td>
</tr>
<tr class="row-odd"><td><p>zbar</p></td>
<td><p>Posterior Probability of Essentiality.</p></td>
</tr>
<tr class="row-even"><td><p>Call</p></td>
<td><p>Essentiality call for the gene. Depends on FDR corrected thresholds. E=Essential U=Uncertain, NE=Non-Essential, S=too short</p></td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line"><br /></div>
<div class="line-block">
<div class="line">Note: Technically, Bayesian models are used to calculate posterior
probabilities, not p-values (which is a concept associated with the
frequentist framework). However, we have implemented a method for
computing the approximate false-discovery rate (FDR) that serves a
similar purpose. This determines a threshold for significance on the
posterior probabilities that is corrected for multiple tests. The
actual thresholds used are reported in the headers of the output file
(and are near 1 for essentials and near 0 for non-essentials). There
can be many genes that score between the two thresholds (t1 &lt; zbar &lt;
t2). This reflects intrinsic uncertainty associated with either low
read counts, sparse insertion density, or small genes. If the
insertion_density is too low (&lt; ~30%), the method may not work as
well, and might indicate an unusually large number of Uncertain or
Essential genes.</div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="run-time">
<h2>Run-time<a class="headerlink" href="#run-time" title="Permalink to this headline"></a></h2>
<p>The Gumbel method takes on the order of 10 minutes for 10,000 samples.
Run-time is linearly proportional to the ‘samples’ parameter, or length
of MH sampling trajectory. Other notes: Gumbel can be run on multiple
replicates; replicate datasets will be automatically merged.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="transit-sectionend docutils" />
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="transit_features.html" class="btn btn-neutral float-left" title="Quality Control" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="method_griffin.html" class="btn btn-neutral float-right" title="Griffin" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2015, Michael A. DeJesus &amp; Thomas R. Ioerger.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>