<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>TRANSIT Overview &mdash; TRANSIT v3.2.4 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/default.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="transit_install.html" />
    <link rel="prev" title="Welcome to TRANSIT’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> TRANSIT
            <img src="_static/transit_logo.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                v3.2.4
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">TRANSIT MANUAL</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">TRANSIT Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#developers">Developers</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="transit_install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_running.html">Running TRANSIT</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_features.html">Quality Control</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">ANALYSIS METHODS</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="method_gumbel.html">Gumbel</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_griffin.html">Griffin</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_tn5gaps.html">Tn5Gaps</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_HMM.html">HMM</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_resampling.html">Resampling</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_Utest.html">Mann-Whitney U-test (utest)</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_GI.html">Genetic Interactions</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_anova.html">ANOVA</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_zinb.html">ZINB</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_normalization.html">Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_pathway_enrichment.html">Pathway Enrichment Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_corrplot.html">corrplot</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_heatmap.html">heatmap</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_ttnfitness.html">TTN-Fitness</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">TRANSIT Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit_interactions_tutorial.html">Tutorial: Genetic Interactions Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_normalization_tutorial.html">Tutorial: Normalize datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_export_tutorial.html">Tutorial: Export datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="transit_console_cheatsheet.html">Console Mode Cheat-Sheet</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">TPP Manual</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tpp.html">TPP Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#running-tpp">Running TPP</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#mapping-to-genomes-with-multiple-contigs">Mapping to Genomes with Multiple Contigs</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#overview-of-data-processing-procedure">Overview of Data Processing Procedure</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#statistics">Statistics</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Code Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit.html">transit package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">TRANSIT</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>TRANSIT Overview</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/transit_overview.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="transit-overview">
<h1>TRANSIT Overview<a class="headerlink" href="#transit-overview" title="Permalink to this headline"></a></h1>
<p>The analysis methods in Transit are also described in this <a class="reference external" href="https://orca1.tamu.edu/essentiality/transit/transit-manual.pdf">PDF manual</a> .</p>
<ul class="simple">
<li><p>This is a software that can be used to analyze Tn-Seq datasets. It includes various statistical calculations of essentiality of genes or genomic regions (including conditional essentiality between 2 conditions). These methods were developed and tested as a collaboration between the Sassetti lab (UMass) and the Ioerger lab (Texas A&amp;M) <a class="reference internal" href="#dejesus2015transit" id="id1"><span>[DeJesus2015TRANSIT]</span></a>.</p></li>
</ul>
<a class="reference internal image-reference" href="_images/transit_interface.png"><img alt="_images/transit_interface.png" class="align-center" src="_images/transit_interface.png" style="width: 600px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>TRANSIT is capable of analyzing TnSeq libraries constructed with Himar1 or <a class="reference internal" href="transit_running.html#tn5-main-overview"><span class="std std-ref">Tn5</span></a> datasets.</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>TRANSIT assumes you have already done pre-processing of raw sequencing files (.fastq) and extracted read counts into a <a class="reference external" href="http://genome.ucsc.edu/goldenpath/help/wiggle.html">.wig formatted file</a>. The .wig file should contain the counts at all sites where an insertion could take place (including sites with no reads). For Himar1 datasets this is all TA sites in the genome. For <a class="reference internal" href="transit_running.html#tn5-main-overview"><span class="std std-ref">Tn5</span></a> datasets this would be all nucleotides in the genome.</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>Note that while refer to “read-counts” throughout the documentation, the <a class="reference external" href="http://www.springer.com/biomed/human+genetics/book/978-1-4939-2397-7">current Himar1 protocol</a> utilizes internal barcodes that can be used to reduce raw read counts to unique template counts, and this this is the intended input to TRANSIT from Himar1 datasets.</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>There are various methods available for pre-processing (converting .fastq files to .wig files). You might have your own scripts (if so, massage the data into .wig format), or you might get the scripts used in the Sassetti lab. For convenience, we are including a separate tool called <a class="reference external" href="http://saclab.tamu.edu/tom/TPP.html">TPP</a> (Tn-Seq Pre-Processor) with this distribution that encodes the way we process .fastq files in the Ioerger lab. It’s a complicated process with many steps (removing transposon prefixes of reads, mapping into genome, identifying barcodes and reducing read counts to template counts).</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>Most of the analysis methods in TRANSIT require an <strong>annotation</strong> to know the gene coordinates and names. This is the top file input in the GUI window. The annotation has to be in a somewhat non-standard format called a “.prot_table”. If you know what you are doing, it is easy to convert annotations for other organisms into .prot_table format. But for convenience, we are distributing the prot_tables for 3 common versions of the H37Rv genome: H37Rv.prot_table (NC_000962.2, from Stewart Cole), H37RvMA2.prot_table (sequenced version from the Sassetti lab), and H37RvBD.prot_table (sequenced by the Broad Institute). All of these are slightly different, and it is <strong>critical</strong> that you use the same annotation file as the reference genome sequence used for mapping the reads (during pre-processing).</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>There are three main types of essentiality analyses: individual,
comparative (pairwise), and multi-condition.</p>
<ul>
<li><p>In <em>individual analysis</em>, the goal is to distinguish essential vs. non-essential in a single growth condition, and to assess the statistical significance of these calls. Two methods for this are the <a class="reference internal" href="method_gumbel.html#gumbel"><span class="std std-ref">Gumbel</span></a> method and the <a class="reference internal" href="method_HMM.html#hmm"><span class="std std-ref">HMM</span></a>. They are computationally distinct. The Gumbel method is looking for significant stretches of TA sites lacking insertions, whereas the HMM looks for regions where the mean read count is locally suppressed or increased. The HMM can detect ‘growth-advantaged’ and ‘growth-defect’ regions. The HMM is also a bit more robust on low-density datasets (with insertion density as low as 20-30%). But both methods have their merits and are complementary.</p></li>
</ul>
</li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<blockquote>
<div><ul class="simple">
<li><p>For <em>comparative analysis</em>, the goal is to determine if the sum of read counts differs significantly between two conditions, for which TRANSIT uses <a class="reference internal" href="method_resampling.html#resampling"><span class="std std-ref">resampling</span></a> (a non-parameteric test analogous to a permutation test). Hence this can be used to identify conditionally essential regions and quantify the statistical significance.  A rank-based Mann-Whitney U-test is also available.</p></li>
</ul>
</div></blockquote>
<div class="line-block">
<div class="line"><br /></div>
</div>
<blockquote>
<div><ul class="simple">
<li><p>For <em>multi-condition analysis</em>, there are two methods for determining whether insertion counts in a gene vary significantly across conditions: <a class="reference internal" href="#zinb"><span class="std std-ref">ZINB</span></a> (Zero-Inflated Negative Binomial) regression, and <a class="reference internal" href="method_anova.html#anova"><span class="std std-ref">ANOVA</span></a>.  In general, we find that ZINB finds more (and better) hits than ANOVA (and even out-performs resampling, for cases with 2 conditions).  Futhermore, ZINB can incorporate additional covariates.</p></li>
</ul>
</div></blockquote>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>TRANSIT has been designed to handle multiple replicates. If you have two or more replicate dataset of the same library selected in the same condition, you can provide them, and more of the computational methods will do something reasonable with them.</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>For those methods that generate p-values, we often also calculate adjusted p-value (or ‘q-values’) which are corrected for multiple tests typically the Benjamini-Hochberg procedure. A typical threshold for significance would be q&lt;0.05 (not p&lt;0.05).</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>It is important to understand the GUI model that TRANSIT uses It allows you to load up datasets (.wig files), select them, choose an analysis method, set parameters, and start the computation. It will generate <strong>output files</strong> in your local directory with the results. These files can then be loaded into the interface and browser with custom displays and graphs. The interface has 3 main windows or sections: ‘Control Samples’, ‘Experimental Samples’, ‘Results Files.’ The first two are for loading input files (‘Control Samples’ would be like replicate datasets from a reference condition, like in vitro, rich media, etc.; ‘Experimental Samples’ would be where you would load replicates for a comparative conditions, like in vivo, or minimal media, or low-iron, etc.) The ‘Results Files’ section is initially empty, but after a computation finishes, it will automatically be populated with the corresponding output file. See the ‘Tutorial’ section below in this documentation for an illustration of the overall process for a typical work-flow.</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>TRANSIT incorporates many interesting ways of looking at your data.</p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
<blockquote>
<div><ul class="simple">
<li><p>Track view shows you a visual representation of the read counts at each site at a locus of interest (for selected datasets) somewhat like IGV.</p></li>
</ul>
</div></blockquote>
<a class="reference internal image-reference" href="_images/transit_dataset_track_view.png"><img alt="_images/transit_dataset_track_view.png" class="align-center" src="_images/transit_dataset_track_view.png" style="width: 600px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
<blockquote>
<div><ul class="simple">
<li><p>Scatter plots can show the correlation of counts between 2 datasets.</p></li>
</ul>
</div></blockquote>
<a class="reference internal image-reference" href="_images/transit_dataset_scatter_graph.png"><img alt="_images/transit_dataset_scatter_graph.png" class="align-center" src="_images/transit_dataset_scatter_graph.png" style="width: 600px;" /></a>
<div class="line-block">
<div class="line">+ Volcano plots can be used to visualize the results of resampling and assess the distribution between over- and under-represented genes in condition B vs. condition A. In addition you can look at histogram of the resampling distributions for each gene.</div>
</div>
<a class="reference internal image-reference" href="_images/transit_result_volcano_graph.png"><img alt="_images/transit_result_volcano_graph.png" class="align-center" src="_images/transit_result_volcano_graph.png" style="width: 600px;" /></a>
<a class="reference internal image-reference" href="_images/transit_resampling_histogram_graph.png"><img alt="_images/transit_resampling_histogram_graph.png" class="align-center" src="_images/transit_resampling_histogram_graph.png" style="width: 600px;" /></a>
<ul class="simple">
<li><p>Most of the methods take a few minutes to run. (it depends on parameters, CPU clock speed, etc., but the point is, a) these calculations are complex and not instaneous, but b) we have tried to implement it so that they don’t take hours)</p></li>
<li><p>Note: in this version of TRANSIT, most of the methods are oriented toward gene-level analysis. There are methods for analyzing essentiality of arbitrary genomic regions (e.g. sliding windows, HMMs…). We plan to incorporate some of these in future versions.</p></li>
</ul>
<section id="developers">
<h2>Developers<a class="headerlink" href="#developers" title="Permalink to this headline"></a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 11%" />
<col style="width: 69%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Time Active</p></th>
<th class="head"><p>Contact Information</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Thomas R. Ioerger</p></td>
<td><p>2015-Present</p></td>
<td><p><a class="reference external" href="http://faculty.cs.tamu.edu/ioerger/">http://faculty.cs.tamu.edu/ioerger/</a></p></td>
</tr>
<tr class="row-odd"><td><p>Michael A. DeJesus</p></td>
<td><p>2015-2018</p></td>
<td><p><a class="reference external" href="http://mad-lab.org">http://mad-lab.org</a></p></td>
</tr>
<tr class="row-even"><td><p>Chaitra Ambadipudi</p></td>
<td><p>2015</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Eric Nelson</p></td>
<td><p>2016</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Siddharth Subramaniyam</p></td>
<td><p>2018</p></td>
<td></td>
</tr>
</tbody>
</table>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline"></a></h2>
<p>If you use TRANSIT, please cite the following reference:</p>
<dl class="citation">
<dt class="label" id="dejesus2015transit"><span class="brackets"><a class="fn-backref" href="#id1">DeJesus2015TRANSIT</a></span></dt>
<dd><p><a class="reference external" href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004401">DeJesus, M.A., Ambadipudi, C., Baker, R., Sassetti, C., and Ioerger, T.R. (2015). TRANSIT - a Software Tool for Himar1 TnSeq Analysis. PLOS Computational Biology, 11(10):e1004401</a></p>
</dd>
</dl>
<p>Development of TRANSIT is funded by the National Institutes of Health (www.nih.gov/) grant U19 AI107774.</p>
<p>Other references for methods utilized by TRANSIT:</p>
<dl class="citation">
<dt class="label" id="dejesus2013"><span class="brackets">DeJesus2013</span></dt>
<dd><p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23361328">DeJesus, M.A., Zhang, Y.J., Sassettti, C.M., Rubin, E.J.,
Sacchettini, J.C., and Ioerger, T.R. (2013). Bayesian analysis of gene essentiality based on sequencing of transposon insertion libraries. Bioinformatics, 29(6):695-703.</a></p>
</dd>
<dt class="label" id="dejesus2013hmm"><span class="brackets">DeJesus2013HMM</span></dt>
<dd><p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24103077">DeJesus, M.A., Ioerger, T.R. A Hidden Markov Model for identifying essential and growth-defect regions in bacterial genomes from transposon insertion sequencing data. BMC Bioinformatics. 2013. 14:303</a></p>
</dd>
<dt class="label" id="dejesus2014"><span class="brackets">DeJesus2014</span></dt>
<dd><p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/26357081">DeJesus, M.A. and Ioerger, T.R. (2014). Capturing uncertainty by modeling local transposon insertion frequencies improves discrimination of essential genes. IEEE Transactions on Computational Biology and Bioinformatics, 12(1):92-102.</a></p>
</dd>
<dt class="label" id="dejesus2016"><span class="brackets">DeJesus2016</span></dt>
<dd><p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/26932272">DeJesus, M.A. and Ioerger, T.R. (2016). Normalization of transposon-mutant library sequencing datasets to improve identification of conditionally essential genes. Journal of Bioinformatics and Computational Biology, 14(3):1642004</a></p>
</dd>
<dt class="label" id="dejesus2017nar"><span class="brackets">DeJesus2017NAR</span></dt>
<dd><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/28334803">DeJesus, M.A., Nambi, S., Smith, C.M., Baker, R.E., Sassetti, C.M., Ioerger, T.R. Statistical analysis of genetic interactions in Tn-Seq data.  Nucleic Acids Research. 2017. 45(11):e93. doi: 10.1093/nar/gkx128.</a></p>
</dd>
<dt class="label" id="zinb"><span class="brackets">ZINB</span></dt>
<dd><p><a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-3156-z">Subramaniyam S, DeJesus MA, Zaveri A, Smith CM, Baker RE, Ehrt S, Schnappinger D, Sassetti CM, Ioerger TR. (2019).  Statistical analysis of variability in TnSeq data across conditions using Zero-Inflated Negative Binomial regression. *BMC Bioinformatics*. 2019 Nov 21;20(1):603. doi: 10.1186/s12859-019-3156-z.</a></p>
</dd>
<dt class="label" id="choudhery2021"><span class="brackets">Choudhery2021</span></dt>
<dd><p><a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/34665010/">Choudhery S, Brown AJ, Akusobi C, Rubin EJ, Sassetti CM, Ioerger TR. Modeling Site-Specific Nucleotide Biases Affecting Himar1 Transposon Insertion Frequencies in TnSeq Data Sets. *mSystems*. 2021 Oct 26;6(5):e0087621. doi: 10.1128/mSystems.00876-21.</a></p>
</dd>
</dl>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to TRANSIT’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="transit_install.html" class="btn btn-neutral float-right" title="Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2015, Michael A. DeJesus &amp; Thomas R. Ioerger.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>