<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pytransit.tnseq_tools &mdash; TRANSIT v3.2.3 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/default.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> TRANSIT
            <img src="../../_static/transit_logo.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                v3.2.3
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">TRANSIT MANUAL</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../transit_overview.html">TRANSIT Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../transit_install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../transit_running.html">Running TRANSIT</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../transit_features.html">Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../transit_methods.html">Analysis Methods (Expand Me First!)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../transit_console_cheatsheet.html">Console Mode Cheat-Sheet</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">TRANSIT Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../transit_interactions_tutorial.html">Tutorial: Genetic Interactions Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../transit_normalization_tutorial.html">Tutorial: Normalize datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../transit_export_tutorial.html">Tutorial: Export datasets</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">TPP Manual</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tpp.html">TPP Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tpp.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tpp.html#running-tpp">Running TPP</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tpp.html#mapping-to-genomes-with-multiple-contigs">Mapping to Genomes with Multiple Contigs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tpp.html#overview-of-data-processing-procedure">Overview of Data Processing Procedure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tpp.html#statistics">Statistics</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Code Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../transit.html">transit package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">TRANSIT</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>pytransit.tnseq_tools</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pytransit.tnseq_tools</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">scipy.stats</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">total_ordering</span>


<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">pytransit</span> <span class="kn">import</span> <span class="n">norm_tools</span>
    <span class="n">noNorm</span> <span class="o">=</span> <span class="kc">False</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">noNorm</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Problem importing the norm_tools.py module. Read-counts will not be normalized. Some functions may not work.&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="rv_siteindexes_map"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.rv_siteindexes_map">[docs]</a><span class="k">def</span> <span class="nf">rv_siteindexes_map</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">TASiteindexMap</span><span class="p">,</span> <span class="n">nterm</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">cterm</span><span class="o">=</span><span class="mf">0.0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    ([Gene], {TAsite: Siteindex}) -&gt; {Rv: Siteindex}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">RvSiteindexesMap</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genes</span><span class="p">):</span>
        <span class="n">siteindexes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span> <span class="k">else</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">3</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="mi">3</span> <span class="k">if</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span> <span class="k">else</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">co</span> <span class="o">=</span> <span class="n">i</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">co</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="p">)</span> <span class="o">&lt;</span> <span class="p">(</span><span class="n">nterm</span><span class="o">/</span><span class="mf">100.0</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">co</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="p">)</span> <span class="o">&gt;</span> <span class="p">((</span><span class="mi">100</span> <span class="o">-</span> <span class="n">cterm</span><span class="p">)</span><span class="o">/</span><span class="mf">100.0</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">co</span> <span class="ow">in</span> <span class="n">TASiteindexMap</span><span class="p">:</span> <span class="n">siteindexes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">TASiteindexMap</span><span class="p">[</span><span class="n">co</span><span class="p">])</span>
        <span class="n">RvSiteindexesMap</span><span class="p">[</span><span class="n">gene</span><span class="p">[</span><span class="s2">&quot;rv&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">siteindexes</span>
    <span class="k">return</span> <span class="n">RvSiteindexesMap</span></div>

<span class="c1"># format:</span>
<span class="c1">#   header lines (prefixed by &#39;#&#39;), followed by lines with counts</span>
<span class="c1">#   counts lines contain the following columns: TA coord, counts, other info like gene/annotation</span>
<span class="c1">#   for each column of counts, there must be a header line prefixed by &quot;#File: &quot; and then an id or filename</span>

<div class="viewcode-block" id="read_combined_wig"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.read_combined_wig">[docs]</a><span class="k">def</span> <span class="nf">read_combined_wig</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read the combined wig-file generated by Transit</span>
<span class="sd">        :: Filename -&gt; Tuple([Site], [WigData], [Filename])</span>
<span class="sd">        Site :: Integer</span>
<span class="sd">        WigData :: [Number]</span>
<span class="sd">        Filename :: String</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sites</span><span class="p">,</span><span class="n">countsByWig</span><span class="p">,</span><span class="n">files</span> <span class="o">=</span> <span class="p">[],[],[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#File: &quot;</span><span class="p">):</span>
                <span class="n">files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()[</span><span class="mi">7</span><span class="p">:])</span> <span class="c1"># allows for spaces in filenames</span>
    <span class="n">countsByWig</span> <span class="o">=</span> <span class="p">[[]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">files</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;#&#39;</span><span class="p">:</span> <span class="k">continue</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">files</span><span class="p">)]</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[:</span><span class="mi">1</span><span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">files</span><span class="p">)]</span> <span class="c1"># additional columns at end could contain gene info</span>
        <span class="c1"># Read in position as int, and readcounts as float</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">t_iv</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">t_iv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">if</span> <span class="n">t_iv</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="nb">float</span><span class="p">(</span><span class="n">t_iv</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cols</span><span class="p">)))</span>
        <span class="n">position</span><span class="p">,</span> <span class="n">wigCounts</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
        <span class="n">sites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">position</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">wigCounts</span><span class="p">):</span>
            <span class="n">countsByWig</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sites</span><span class="p">),</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">countsByWig</span><span class="p">),</span> <span class="n">files</span><span class="p">)</span></div>

<div class="viewcode-block" id="read_samples_metadata"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.read_samples_metadata">[docs]</a><span class="k">def</span> <span class="nf">read_samples_metadata</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">,</span> <span class="n">covarsToRead</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">interactionsToRead</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">condition_name</span><span class="o">=</span><span class="s2">&quot;Condition&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">      Filename -&gt; ConditionMap</span>
<span class="sd">      ConditionMap :: {Filename: Condition}, [{Filename: Covar}], [{Filename: Interaction}]</span>
<span class="sd">      Condition :: String</span>
<span class="sd">      Covar :: String</span>
<span class="sd">      Interaction :: String</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">wigFiles</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">conditionsByFile</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">covariatesByFileList</span> <span class="o">=</span> <span class="p">[{}</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">covarsToRead</span><span class="p">))]</span>
    <span class="n">interactionsByFileList</span> <span class="o">=</span> <span class="p">[{}</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">interactionsToRead</span><span class="p">))]</span>
    <span class="n">headersToRead</span> <span class="o">=</span> <span class="p">[</span><span class="n">condition_name</span><span class="o">.</span><span class="n">lower</span><span class="p">(),</span> <span class="s2">&quot;filename&quot;</span><span class="p">]</span>
    <span class="n">orderingMetadata</span> <span class="o">=</span> <span class="p">{</span> <span class="s1">&#39;condition&#39;</span><span class="p">:</span> <span class="p">[],</span> <span class="s1">&#39;interaction&#39;</span><span class="p">:</span> <span class="p">[]</span> <span class="p">}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">mfile</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="n">mfile</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="n">headIndexes</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span>
                <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">headersToRead</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>
                <span class="k">if</span> <span class="n">c</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="n">h</span><span class="o">.</span><span class="n">lower</span><span class="p">()]</span>
        <span class="n">covarIndexes</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span>
                <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">covarsToRead</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>
                <span class="k">if</span> <span class="n">c</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="n">h</span><span class="o">.</span><span class="n">lower</span><span class="p">()]</span>
        <span class="n">interactionIndexes</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span>
                <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">interactionsToRead</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>
                <span class="k">if</span> <span class="n">c</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="n">h</span><span class="o">.</span><span class="n">lower</span><span class="p">()]</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;#&#39;</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="p">[</span><span class="n">condition</span><span class="p">,</span> <span class="n">wfile</span><span class="p">]</span> <span class="o">=</span> <span class="n">vals</span><span class="p">[</span><span class="n">headIndexes</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">vals</span><span class="p">[</span><span class="n">headIndexes</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
            <span class="n">conditionsByFile</span><span class="p">[</span><span class="n">wfile</span><span class="p">]</span> <span class="o">=</span> <span class="n">condition</span>
            <span class="n">orderingMetadata</span><span class="p">[</span><span class="s1">&#39;condition&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">condition</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">covarsToRead</span><span class="p">):</span>
                <span class="n">covariatesByFileList</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">wfile</span><span class="p">]</span> <span class="o">=</span> <span class="n">vals</span><span class="p">[</span><span class="n">covarIndexes</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">interactionsToRead</span><span class="p">):</span>
                <span class="n">interactionsByFileList</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">wfile</span><span class="p">]</span> <span class="o">=</span> <span class="n">vals</span><span class="p">[</span><span class="n">interactionIndexes</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>

                <span class="c1"># This makes sense only if there is only 1 interaction variable</span>
                <span class="c1"># For multiple interaction vars, may have to rethink ordering.</span>
                <span class="n">orderingMetadata</span><span class="p">[</span><span class="s1">&#39;interaction&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vals</span><span class="p">[</span><span class="n">interactionIndexes</span><span class="p">[</span><span class="n">i</span><span class="p">]])</span>

    <span class="k">return</span> <span class="n">conditionsByFile</span><span class="p">,</span> <span class="n">covariatesByFileList</span><span class="p">,</span> <span class="n">interactionsByFileList</span><span class="p">,</span> <span class="n">orderingMetadata</span></div>

<div class="viewcode-block" id="read_genes"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.read_genes">[docs]</a><span class="k">def</span> <span class="nf">read_genes</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span><span class="n">descriptions</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">      (Filename, Options) -&gt; [Gene]</span>
<span class="sd">      Gene :: {start, end, rv, gene, strand}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
        <span class="n">w</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">w</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">w</span><span class="p">[</span><span class="mi">2</span><span class="p">]),</span>
                <span class="s2">&quot;rv&quot;</span><span class="p">:</span> <span class="n">w</span><span class="p">[</span><span class="mi">8</span><span class="p">],</span>
                <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">w</span><span class="p">[</span><span class="mi">7</span><span class="p">],</span>
                <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">w</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
                <span class="p">}</span>
        <span class="k">if</span> <span class="n">descriptions</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span> <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">w</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genes</span></div>

<div class="viewcode-block" id="Gene"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene">[docs]</a><span class="nd">@total_ordering</span>
<span class="k">class</span> <span class="nc">Gene</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Class defining a gene with useful attributes for TnSeq analysis.</span>

<span class="sd">    This class helps define a &quot;gene&quot; with attributes that facilitate TnSeq</span>
<span class="sd">    analysis. Here &quot;gene&quot; can be defined to be any genomic region. The Genes</span>
<span class="sd">    class (with an s) can be used to define list of Gene objects with more</span>
<span class="sd">    useful operations on the &quot;genome&quot; level.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        orf: A string defining the ID of the gene.</span>
<span class="sd">        name: A string with the human readable name of the gene.</span>
<span class="sd">        desc: A string with the description of the gene.</span>
<span class="sd">        reads: List of lists of read-counts in possible site replicate dataset.</span>
<span class="sd">        position: List of coordinates of the possible sites.</span>
<span class="sd">        start: An integer defining the start coordinate for the gene.</span>
<span class="sd">        end: An integer defining the end coordinate for the gene.</span>
<span class="sd">        strand: A string defining the strand of the gene.</span>


<span class="sd">    :Example:</span>

<span class="sd">        &gt;&gt;&gt; import pytransit.tnseq_tools as tnseq_tools</span>
<span class="sd">        &gt;&gt;&gt; G = tnseq_tools.Gene(&quot;Rv0001&quot;, &quot;dnaA&quot;, &quot;DNA Replication A&quot;, [[0,0,0,0,1,3,0,1]],  [1,21,32,37,45,58,66,130], strand=&quot;+&quot; )</span>
<span class="sd">        &gt;&gt;&gt; print(G)</span>
<span class="sd">        Rv0001  (dnaA)  k=3 n=8 r=4 theta=0.37500</span>
<span class="sd">        &gt;&gt;&gt; print(G.phi())</span>
<span class="sd">        0.625</span>
<span class="sd">        &gt;&gt;&gt; print(G.tosses)</span>
<span class="sd">        array([ 0.,  0.,  0.,  0.,  1.,  1.,  0.,  1.])</span>

<span class="sd">        .. seealso:: :class:`Genes`</span>
<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">orf</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">desc</span><span class="p">,</span> <span class="n">reads</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Initializes the Gene object.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            orf (str): A string defining the ID of the gene.</span>
<span class="sd">            name (str): A string with the human readable name of the gene.</span>
<span class="sd">            desc (str): A string with the description of the gene.</span>
<span class="sd">            reads (list): List of lists of read-counts in possible site replicate dataset.</span>
<span class="sd">            position (list): List of coordinates of the possible sites.</span>
<span class="sd">            start (int): An integer defining the start coordinate for the gene.</span>
<span class="sd">            end (int): An integer defining the end coordinate for the gene.</span>
<span class="sd">            strand (str): A string defining the strand of the gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Gene: Object of the Gene class with the defined attributes.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">orf</span> <span class="o">=</span> <span class="n">orf</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">desc</span> <span class="o">=</span> <span class="n">desc</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">start</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="n">end</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="n">strand</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tosses</span> <span class="o">=</span> <span class="n">tossify</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">runs</span> <span class="o">=</span> <span class="n">runs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tosses</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">orf</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">tosses</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">e</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tosses</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tosses</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">r</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">runs</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">orf</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">tosses</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">runs</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">e</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">s</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gap_span</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">t</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_span</span><span class="p">()</span>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.__getitem__"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.__getitem__">[docs]</a>    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return read-counts at position i.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            i (int): integer of the index of the desired site.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: Reads at position i.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.__str__"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.__str__">[docs]</a>    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a string representation of the object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Human readable string with some of the attributes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="se">\t</span><span class="s2">(</span><span class="si">%s</span><span class="s2">)</span><span class="se">\t</span><span class="s2">k=</span><span class="si">%d</span><span class="se">\t</span><span class="s2">n=</span><span class="si">%d</span><span class="se">\t</span><span class="s2">r=</span><span class="si">%d</span><span class="se">\t</span><span class="s2">theta=</span><span class="si">%1.5f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">orf</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">r</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">theta</span><span class="p">())</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.__eq__"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.__eq__">[docs]</a>    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Compares against other gene object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool: True if the gene objects have same orf id.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">orf</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">orf</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.__lt__"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.__lt__">[docs]</a>    <span class="k">def</span> <span class="fm">__lt__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Compares against other gene object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool: True if the gene object id is less than the other.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">orf</span> <span class="o">&lt;</span> <span class="n">other</span><span class="o">.</span><span class="n">orf</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.get_gap_span"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.get_gap_span">[docs]</a>    <span class="k">def</span> <span class="nf">get_gap_span</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the span of the maxrun of the gene (i.e. number of nucleotides).</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: Number of nucleotides spanned by the max run.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">r</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">return</span> <span class="mi">0</span>
            <span class="n">index</span> <span class="o">=</span> <span class="n">runindex</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">runs</span><span class="p">)</span>
            <span class="c1">#maxii = numpy.argmax(self.runs)</span>
            <span class="n">maxii</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">runs</span> <span class="o">==</span> <span class="n">numpy</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">runs</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">runstart</span> <span class="o">=</span> <span class="n">index</span><span class="p">[</span><span class="n">maxii</span><span class="p">]</span>
            <span class="n">runend</span> <span class="o">=</span> <span class="n">runstart</span> <span class="o">+</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">runs</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">runend</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">runstart</span><span class="p">]</span> <span class="o">+</span> <span class="mi">2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">0</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.get_gene_span"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.get_gene_span">[docs]</a>    <span class="k">def</span> <span class="nf">get_gene_span</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the number of nucleotides spanned by the gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: Number of nucleotides spanned by the gene&#39;s sites.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">2</span>
        <span class="k">return</span> <span class="mi">0</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.theta"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.theta">[docs]</a>    <span class="k">def</span> <span class="nf">theta</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return the insertion density (&quot;theta&quot;) for the gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Density of the gene (i.e. k/n )</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">n</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">k</span><span class="p">)</span><span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">n</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.phi"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.phi">[docs]</a>    <span class="k">def</span> <span class="nf">phi</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return the non-insertion density (&quot;phi&quot;) for the gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Non-insertion density  (i.e. 1 - theta)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">theta</span><span class="p">()</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Gene.total_reads"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Gene.total_reads">[docs]</a>    <span class="k">def</span> <span class="nf">total_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return the total reads for the gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Total sum of read-counts.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reads</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span></div></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes">[docs]</a><span class="k">class</span> <span class="nc">Genes</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Class defining a list of Gene objects with useful attributes for TnSeq</span>
<span class="sd">    analysis.</span>

<span class="sd">    This class helps define a list of Gene objects with attributes that</span>
<span class="sd">    facilitate TnSeq analysis. Includes methods that calculate useful statistics</span>
<span class="sd">    and even rudamentary analysis of essentiality.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        wigList: List of paths to datasets in .wig format.</span>
<span class="sd">        protTable: String with path to annotation in .prot_table format.</span>
<span class="sd">        norm: String with the normalization used/</span>
<span class="sd">        reps: String with information on how replicates were handled.</span>
<span class="sd">        minread: Integer with the minimum magnitude of read-count considered.</span>
<span class="sd">        ignoreCodon: Boolean defining whether to ignore the start/stop codon.</span>
<span class="sd">        nterm: Float number of the fraction of the N-terminus to ignore.</span>
<span class="sd">        cterm: Float number of the fraction of the C-terminus to ignore.</span>
<span class="sd">        include_nc: Boolean determining whether to include non-coding areas.</span>
<span class="sd">        orf2index: Dictionary of orf id to index in the genes list.</span>
<span class="sd">        genes: List of the Gene objects.</span>


<span class="sd">    :Example:</span>

<span class="sd">        &gt;&gt;&gt; import pytransit.tnseq_tools as tnseq_tools</span>
<span class="sd">        &gt;&gt;&gt; G = tnseq_tools.Genes([&quot;transit/data/glycerol_H37Rv_rep1.wig&quot;, &quot;transit/data/glycerol_H37Rv_rep2.wig&quot;], &quot;transit/genomes/H37Rv.prot_table&quot;, norm=&quot;TTR&quot;)</span>
<span class="sd">        &gt;&gt;&gt; print(G)</span>
<span class="sd">        Genes Object (N=3990)</span>
<span class="sd">        &gt;&gt;&gt; print(G.global_theta())</span>
<span class="sd">        0.40853707222816626</span>
<span class="sd">        &gt;&gt;&gt; print(G[&quot;Rv0001&quot;]   # Lookup like dictionary)</span>
<span class="sd">        Rv0001  (dnaA)  k=0 n=31    r=31    theta=0.00000</span>
<span class="sd">        &gt;&gt;&gt; print(G[2]          # Lookup like list)</span>
<span class="sd">        Rv0003  (recF)  k=5 n=35    r=14    theta=0.14286</span>
<span class="sd">        &gt;&gt;&gt; print(G[2].reads)</span>
<span class="sd">        [[  62.            0.            0.            0.            0.            0.</span>
<span class="sd">         0.            0.            0.            0.            0.            0.</span>
<span class="sd">         0.            0.           63.            0.            0.           13.</span>
<span class="sd">        46.            0.            1.            0.            0.            0.</span>
<span class="sd">         0.            0.            0.            0.            0.            0.</span>
<span class="sd">         0.            0.            0.            0.            0.        ]</span>
<span class="sd">         [   3.14314432   67.26328843    0.            0.            0.            0.</span>
<span class="sd">         0.            0.            0.           35.20321637    0.            0.</span>
<span class="sd">         0.            0.           30.80281433    0.          101.20924707</span>
<span class="sd">         0.           23.25926796    0.           16.97297932    8.17217523</span>
<span class="sd">         0.            0.            2.51451546    3.77177318    0.62862886</span>
<span class="sd">         0.            0.           69.14917502    0.            0.            0.</span>
<span class="sd">         0.            0.        ]]</span>


<span class="sd">        .. seealso:: :class:`Gene`</span>
<span class="sd">        &quot;&quot;&quot;</span>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.__getitem__"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.__getitem__">[docs]</a>    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Defines __getitem__ method so that it works as dictionary and list.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            i (int): Integer or string defining index or orf ID desired.</span>
<span class="sd">        Returns:</span>
<span class="sd">            Gene: A gene with the index or ID equal to i.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">return</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">orf2index</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.__contains__"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.__contains__">[docs]</a>    <span class="k">def</span> <span class="fm">__contains__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">item</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Defines __contains__ to check if gene exists in the list.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            item (str): String with the id of the gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            bool: Boolean with True if item is in the list.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">item</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">orf2index</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.__len__"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.__len__">[docs]</a>    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Defines __len__ returning number of genes.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: Number of genes in the list.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.__str__"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.__str__">[docs]</a>    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Defines __str__ to print(a generic str with the size of the list.)</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Human readable string with number of genes in object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;Genes Object (N=</span><span class="si">%d</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span></div>

<span class="c1">#</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">wigList</span><span class="p">,</span> <span class="n">annotation</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="s2">&quot;nonorm&quot;</span><span class="p">,</span> <span class="n">reps</span><span class="o">=</span><span class="s2">&quot;All&quot;</span><span class="p">,</span> <span class="n">minread</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ignoreCodon</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">nterm</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">cterm</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">include_nc</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="p">[],</span> <span class="n">position</span><span class="o">=</span><span class="p">[],</span><span class="n">genome</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">transposon</span><span class="o">=</span><span class="s2">&quot;himar1&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Initializes the gene list based on the list of wig files and a prot_table.</span>

<span class="sd">        This class helps define a list of Gene objects with attributes that</span>
<span class="sd">        facilitate TnSeq analysis. Includes methods that calculate useful statistics</span>
<span class="sd">        and even rudamentary analysis of essentiality.</span>

<span class="sd">        Arguments:</span>
<span class="sd">            wigList (list): List of paths to datasets in .wig format.</span>
<span class="sd">            protTable (str): String with path to annotation in .prot_table format.</span>
<span class="sd">            norm (str): String with the normalization used/</span>
<span class="sd">            reps (str): String with information on how replicates were handled.</span>
<span class="sd">            minread (int): Integer with the minimum magnitude of read-count considered.</span>
<span class="sd">            ignoreCodon (bool): Boolean defining whether to ignore the start/stop codon.</span>
<span class="sd">            nterm (float): Float number of the fraction of the N-terminus to ignore.</span>
<span class="sd">            cterm (float): Float number of the fraction of the C-terminus to ignore.</span>
<span class="sd">            include_nc (bool): Boolean determining whether to include non-coding areas.</span>
<span class="sd">            data (list): List of data. Used to define the object without files.</span>
<span class="sd">            position (list): List of position of sites. Used to define the object without files.</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">wigList</span> <span class="o">=</span> <span class="n">wigList</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="o">=</span> <span class="n">annotation</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">norm</span> <span class="o">=</span> <span class="n">norm</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reps</span> <span class="o">=</span> <span class="n">reps</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">minread</span> <span class="o">=</span> <span class="n">minread</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ignoreCodon</span> <span class="o">=</span> <span class="n">ignoreCodon</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nterm</span> <span class="o">=</span> <span class="n">nterm</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cterm</span> <span class="o">=</span> <span class="n">cterm</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">include_nc</span> <span class="o">=</span> <span class="n">include_nc</span>

        <span class="n">isProt</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">filename</span><span class="p">,</span> <span class="n">file_extension</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">file_extension</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;.gff&quot;</span><span class="p">,</span> <span class="s2">&quot;.gff3&quot;</span><span class="p">]:</span>
            <span class="n">isProt</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">orf2index</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">orf2info</span> <span class="o">=</span> <span class="n">get_gene_info</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">numpy</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">transposon</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;himar1&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">genome</span><span class="p">:</span>
                <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span> <span class="o">=</span> <span class="n">get_data</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">wigList</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">genome</span><span class="p">:</span>
                <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span> <span class="o">=</span> <span class="n">get_data_w_genome</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">wigList</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span> <span class="o">=</span> <span class="n">get_data_zero_fill</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">wigList</span><span class="p">)</span>

        <span class="n">ii_min</span> <span class="o">=</span> <span class="n">data</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">minread</span>
        <span class="n">data</span><span class="p">[</span><span class="n">ii_min</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="nb">hash</span> <span class="o">=</span> <span class="n">get_pos_hash</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">noNorm</span><span class="p">:</span>
            <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">factors</span><span class="p">)</span> <span class="o">=</span> <span class="n">norm_tools</span><span class="o">.</span><span class="n">normalize_data</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">norm</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">wigList</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">factors</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">reps</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">combine_replicates</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">reps</span><span class="p">)])</span>

        <span class="n">K</span><span class="p">,</span><span class="n">N</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="n">orf2posindex</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">visited_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">):</span>
            <span class="n">genes_with_coord</span> <span class="o">=</span> <span class="nb">hash</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="p">[])</span>
            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes_with_coord</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">orf2posindex</span><span class="p">:</span> <span class="n">visited_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">orf2posindex</span><span class="p">:</span> <span class="n">orf2posindex</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

                <span class="n">name</span><span class="p">,</span><span class="n">desc</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">,</span><span class="n">strand</span> <span class="o">=</span> <span class="n">orf2info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">])</span>

                <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignoreCodon</span> <span class="ow">and</span> <span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">3</span><span class="p">:</span>
                        <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignoreCodon</span> <span class="ow">and</span> <span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">start</span> <span class="o">+</span> <span class="mi">3</span><span class="p">:</span>
                        <span class="k">continue</span>

                <span class="k">if</span> <span class="p">(</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">-</span><span class="n">start</span><span class="p">)</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="p">)</span> <span class="o">&lt;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">nterm</span><span class="o">/</span><span class="mf">100.0</span><span class="p">):</span>
                    <span class="k">continue</span>

                <span class="k">if</span> <span class="p">(</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">-</span><span class="n">start</span><span class="p">)</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="p">)</span> <span class="o">&gt;</span> <span class="p">((</span><span class="mi">100</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">cterm</span><span class="p">)</span><span class="o">/</span><span class="mf">100.0</span><span class="p">):</span>
                    <span class="k">continue</span>

                <span class="n">orf2posindex</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span> <span class="k">continue</span>
            <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">isProt</span><span class="p">:</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                <span class="n">name</span><span class="p">,</span><span class="n">desc</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">,</span><span class="n">strand</span> <span class="o">=</span> <span class="n">orf2info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">features</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;=&quot;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">,</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">))])</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">features</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span>
                <span class="n">name</span><span class="p">,</span><span class="n">desc</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">,</span><span class="n">strand</span> <span class="o">=</span> <span class="n">orf2info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">])</span>
            <span class="n">posindex</span> <span class="o">=</span> <span class="n">orf2posindex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="p">[])</span>
            <span class="k">if</span> <span class="n">posindex</span><span class="p">:</span>
                <span class="n">pos_start</span> <span class="o">=</span> <span class="n">orf2posindex</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">pos_end</span> <span class="o">=</span> <span class="n">orf2posindex</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">desc</span><span class="p">,</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">pos_start</span><span class="p">:</span><span class="n">pos_end</span><span class="o">+</span><span class="mi">1</span><span class="p">],</span> <span class="n">position</span><span class="p">[</span><span class="n">pos_start</span><span class="p">:</span><span class="n">pos_end</span><span class="o">+</span><span class="mi">1</span><span class="p">],</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">desc</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([[]]),</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([]),</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">orf2index</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="n">count</span>
            <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.local_insertions"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.local_insertions">[docs]</a>    <span class="k">def</span> <span class="nf">local_insertions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns numpy array with the number of insertions, &#39;k&#39;, for each gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            narray: Numpy array with the number of insertions for all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">K</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">K</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">k</span>
        <span class="k">return</span> <span class="n">K</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.local_sites"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.local_sites">[docs]</a>    <span class="k">def</span> <span class="nf">local_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns numpy array with total number of TA sites, &#39;n&#39;, for each gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            narray: Numpy array with the number of sites for all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">N</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">N</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">n</span>
        <span class="k">return</span> <span class="n">N</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.local_runs"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.local_runs">[docs]</a>    <span class="k">def</span> <span class="nf">local_runs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns numpy array with maximum run of non-insertions, &#39;r&#39;, for each gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            narray: Numpy array with the max run of non-insertions for all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">R</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">R</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">r</span>
        <span class="k">return</span> <span class="n">R</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.local_gap_span"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.local_gap_span">[docs]</a>    <span class="k">def</span> <span class="nf">local_gap_span</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns numpy array with the span of nucleotides of the largest gap,</span>
<span class="sd">        &#39;s&#39;, for each gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            narray: Numpy array with the span of gap for all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">S</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">S</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">s</span>
        <span class="k">return</span> <span class="n">S</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.local_gene_span"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.local_gene_span">[docs]</a>    <span class="k">def</span> <span class="nf">local_gene_span</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns numpy array with the span of nucleotides of the gene,</span>
<span class="sd">        &#39;t&#39;, for each gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            narray: Numpy array with the span of gene for all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">T</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">T</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">t</span>
        <span class="k">return</span> <span class="n">T</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.local_reads"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.local_reads">[docs]</a>    <span class="k">def</span> <span class="nf">local_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns numpy array of lists containing the read counts for each gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            narray: Numpy array with the list of reads for all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_reads</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">all_reads</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">reads</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">all_reads</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.local_thetas"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.local_thetas">[docs]</a>    <span class="k">def</span> <span class="nf">local_thetas</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns numpy array of insertion frequencies, &#39;theta&#39;, for each gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            narray: Numpy array with the density for all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">theta</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">theta</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">theta</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">theta</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.local_phis"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.local_phis">[docs]</a>    <span class="k">def</span> <span class="nf">local_phis</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns numpy array of non-insertion frequency, &#39;phi&#39;, for each gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            narray: Numpy array with the complement of density for all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">theta</span><span class="p">()</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.global_insertion"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.global_insertion">[docs]</a>    <span class="k">def</span> <span class="nf">global_insertion</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns total number of insertions, i.e. sum of &#39;k&#39; over all genes.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Total sum of reads across all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">total</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">k</span>
        <span class="k">return</span> <span class="n">total</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.global_sites"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.global_sites">[docs]</a>    <span class="k">def</span> <span class="nf">global_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns total number of sites, i.e. sum of &#39;n&#39; over all genes.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: Total number of sites across all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
            <span class="n">total</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">n</span>
        <span class="k">return</span> <span class="n">total</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.global_run"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.global_run">[docs]</a>    <span class="k">def</span> <span class="nf">global_run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the run assuming all genes were concatenated together.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: Max run across all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">maxrun</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tosses</span><span class="p">())</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.global_reads"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.global_reads">[docs]</a>    <span class="k">def</span> <span class="nf">global_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the reads among the library.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: List of all the data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.global_theta"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.global_theta">[docs]</a>    <span class="k">def</span> <span class="nf">global_theta</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns global insertion frequency, of the library.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Total sites with insertions divided by total sites.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">global_insertion</span><span class="p">())</span><span class="o">/</span><span class="bp">self</span><span class="o">.</span><span class="n">global_sites</span><span class="p">()</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.global_phi"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.global_phi">[docs]</a>    <span class="k">def</span> <span class="nf">global_phi</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns global non-insertion frequency, of the library.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Complement of global theta i.e. 1.0-theta</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">global_theta</span><span class="p">()</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.total_reads"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.total_reads">[docs]</a>    <span class="k">def</span> <span class="nf">total_reads</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns total reads among the library.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Total sum of read-counts accross all genes.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">reads_total</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
            <span class="n">reads_total</span> <span class="o">+=</span> <span class="n">g</span><span class="o">.</span><span class="n">total_reads</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">reads_total</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="Genes.tosses"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.Genes.tosses">[docs]</a>    <span class="k">def</span> <span class="nf">tosses</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns list of bernoulli trials, &#39;tosses&#39;, representing insertions in the gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list: Sites represented as bernoulli trials with insertions as true.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_tosses</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
            <span class="n">all_tosses</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">tosses</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">all_tosses</span></div></div>

<span class="c1">#</span>

<div class="viewcode-block" id="tossify"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.tossify">[docs]</a><span class="k">def</span> <span class="nf">tossify</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reduces the data into Bernoulli trials (or &#39;tosses&#39;) based on whether counts were observed or not.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        data (list): List of numeric data.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: Data represented as bernoulli trials with &gt;0 as true.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">K</span><span class="p">,</span><span class="n">N</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span>
    <span class="n">reduced</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">N</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="runs"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.runs">[docs]</a><span class="k">def</span> <span class="nf">runs</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return list of all the runs of consecutive non-insertions.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        data (list): List of numeric data.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of the length of the runs of non-insertions. Non-zero sites are treated as runs of zero.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">runs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">current_r</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">read</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># If ending a run of zeros</span>
            <span class="k">if</span> <span class="n">current_r</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># If we were in a run, add to list</span>
                <span class="n">runs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_r</span><span class="p">)</span>
            <span class="n">current_r</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">runs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_r</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">current_r</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="c1"># If we ended in a run, add it</span>
    <span class="k">if</span> <span class="n">current_r</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">runs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_r</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">runs</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">runs</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="runindex"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.runindex">[docs]</a><span class="k">def</span> <span class="nf">runindex</span><span class="p">(</span><span class="n">runs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a list of the indexes of the start of the runs; complements runs().</span>

<span class="sd">    Arguments:</span>
<span class="sd">        runs (list): List of numeric data.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of the index of the runs of non-insertions. Non-zero sites are treated as runs of zero.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">index_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">runindex</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">runs</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">r</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">runindex</span> <span class="o">=</span> <span class="n">index</span>
            <span class="n">index</span><span class="o">+=</span><span class="mi">1</span>
        <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">runindex</span> <span class="o">=</span> <span class="n">index</span>
            <span class="n">index</span><span class="o">+=</span><span class="mi">1</span>
        <span class="n">index_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">runindex</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">index_list</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_file_types"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_file_types">[docs]</a><span class="k">def</span> <span class="nf">get_file_types</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns the transposon type (himar1/tn5) of the list of wig files.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        wig_list (list): List of paths to wig files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of transposon type (&quot;himar1&quot; or &quot;tn5&quot;).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">wig_list</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>

    <span class="n">types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;tn5&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">wig_list</span><span class="p">))]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">wig_filename</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">wig_filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">wig_file</span><span class="p">:</span>
            <span class="n">prev_pos</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">wig_file</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="s2">&quot;0123456789&quot;</span><span class="p">:</span> <span class="k">continue</span>
                <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">rd</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">pos</span> <span class="o">!=</span> <span class="n">prev_pos</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">types</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;himar1&#39;</span>
                    <span class="k">break</span>
                <span class="n">prev_pos</span> <span class="o">=</span> <span class="n">pos</span>
    <span class="k">return</span> <span class="n">types</span></div>

<div class="viewcode-block" id="check_wig_includes_zeros"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.check_wig_includes_zeros">[docs]</a><span class="k">def</span> <span class="nf">check_wig_includes_zeros</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns boolean list showing whether the given files include empty sites</span>
<span class="sd">    (zero) or not.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        wig_list (list): List of paths to wig files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of boolean values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">wig_list</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>
    <span class="n">includes</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">wig_list</span><span class="p">))]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">wig_filename</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">wig_filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">wig_file</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">wig_file</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="s2">&quot;0123456789&quot;</span><span class="p">:</span> <span class="k">continue</span>
                <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">rd</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">rd</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">includes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">break</span>
    <span class="k">return</span> <span class="n">includes</span></div>



<span class="c1">#</span>

<div class="viewcode-block" id="get_unknown_file_types"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_unknown_file_types">[docs]</a><span class="k">def</span> <span class="nf">get_unknown_file_types</span><span class="p">(</span><span class="n">wig_list</span><span class="p">,</span> <span class="n">transposons</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; &quot;&quot;&quot;</span>
    <span class="c1">#TODO</span>
    <span class="n">file_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">get_file_types</span><span class="p">(</span><span class="n">wig_list</span><span class="p">))</span>
    <span class="n">method_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">transposons</span><span class="p">)</span>
    <span class="n">extra_types</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">file_types</span> <span class="o">-</span> <span class="n">method_types</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">extra_types</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_data"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_data">[docs]</a><span class="k">def</span> <span class="nf">get_data</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Returns a tuple of (data, position) containing a matrix of raw read-counts</span>
<span class="sd">        , and list of coordinates.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        wig_list (list): List of paths to wig files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: Two lists containing data and positions of the wig files given.</span>

<span class="sd">    :Example:</span>

<span class="sd">        &gt;&gt;&gt; import pytransit.tnseq_tools as tnseq_tools</span>
<span class="sd">        &gt;&gt;&gt; (data, position) = tnseq_tools.get_data([&quot;data/glycerol_H37Rv_rep1.wig&quot;, &quot;data/glycerol_H37Rv_rep2.wig&quot;])</span>
<span class="sd">        &gt;&gt;&gt; print(data)</span>
<span class="sd">        array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],</span>
<span class="sd">               [ 0.,  0.,  0., ...,  0.,  0.,  0.]])</span>

<span class="sd">    .. seealso:: :class:`get_file_types` :class:`combine_replicates` :class:`get_data_zero_fill` :class:`pytransit.norm_tools.normalize_data`</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">K</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">wig_list</span><span class="p">)</span>

    <span class="c1"># If empty just quickly return empty lists</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">wig_list</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">)),</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="p">[])</span>

    <span class="c1"># Check size of all wig file matches</span>
    <span class="n">size_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
        <span class="n">T</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="s2">&quot;0123456789&quot;</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">T</span><span class="o">+=</span><span class="mi">1</span>
        <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>

    <span class="c1"># If it doesn&#39;t match, report an error and quit</span>
    <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">size_list</span><span class="p">)</span> <span class="o">!=</span> <span class="p">(</span><span class="n">T</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">size_list</span><span class="p">)):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error: Not all wig files have the same number of sites.&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;       Make sure all .wig files come from the same strain.&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">()</span>

    <span class="n">data</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">K</span><span class="p">,</span><span class="n">T</span><span class="p">))</span>
    <span class="n">position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
        <span class="n">reads</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">prev_pos</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="s2">&quot;0123456789&quot;</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">rd</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">prev_pos</span> <span class="o">=</span> <span class="n">pos</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">data</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">rd</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Error: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">e</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Make sure that all wig files have the same number of TA sites (i.e. same strain)&quot;</span><span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">()</span>
            <span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">pos</span>
            <span class="n">i</span><span class="o">+=</span><span class="mi">1</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_data_zero_fill"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_data_zero_fill">[docs]</a><span class="k">def</span> <span class="nf">get_data_zero_fill</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Returns a tuple of (data, position) containing a matrix of raw read counts,</span>
<span class="sd">        and list of coordinates. Positions that are missing are filled in as zero.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        wig_list (list): List of paths to wig files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: Two lists containing data and positions of the wig files given.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">K</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">wig_list</span><span class="p">)</span>
    <span class="n">T</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">wig_list</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">)),</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="p">[])</span>

    <span class="c1">#NOTE:  This might be slow as we need to find the last insertion site</span>
    <span class="c1">#       over all the replicates. This might be an area to attempt to optimize.</span>
    <span class="n">last_line</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">wig_name</span> <span class="ow">in</span> <span class="n">wig_list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">wig_name</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="s2">&quot;0123456789&quot;</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">last_line</span> <span class="o">=</span> <span class="n">line</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">last_line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">T</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">T</span><span class="p">,</span><span class="n">pos</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">T</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">)),</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="p">[])</span>

    <span class="n">data</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">K</span><span class="p">,</span><span class="n">T</span><span class="p">))</span>
    <span class="n">position</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">T</span><span class="p">))</span> <span class="o">+</span> <span class="mi">1</span><span class="c1">#numpy.zeros(T)</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
        <span class="n">reads</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="s2">&quot;0123456789&quot;</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">rd</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">prev_pos</span> <span class="o">=</span> <span class="n">pos</span>
            <span class="n">data</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">pos</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">rd</span>
            <span class="n">i</span><span class="o">+=</span><span class="mi">1</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_data_w_genome"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_data_w_genome">[docs]</a><span class="k">def</span> <span class="nf">get_data_w_genome</span><span class="p">(</span><span class="n">wig_list</span><span class="p">,</span> <span class="n">genome</span><span class="p">):</span>

    <span class="n">X</span> <span class="o">=</span> <span class="n">read_genome</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
    <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
    <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">pos2index</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;TA&quot;</span><span class="p">:</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span>
            <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
            <span class="n">pos2index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">count</span>
            <span class="n">count</span> <span class="o">+=</span><span class="mi">1</span>

    <span class="n">positions</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span>
    <span class="n">T</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span>
    <span class="n">K</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">wig_list</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">K</span><span class="p">,</span><span class="n">T</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">wig_list</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="s2">&quot;0123456789&quot;</span><span class="p">:</span> <span class="k">continue</span>
            <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">rd</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">pos2index</span><span class="p">:</span>
                <span class="n">index</span> <span class="o">=</span> <span class="n">pos2index</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span>
                <span class="n">data</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">rd</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: Coordinate </span><span class="si">%d</span><span class="s2"> did not match a TA site in the genome. Ignoring counts.&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">pos</span><span class="p">))</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">positions</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="combine_replicates"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.combine_replicates">[docs]</a><span class="k">def</span> <span class="nf">combine_replicates</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;Sum&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns list of data merged together.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        data (list): List of numeric (replicate) data to be merged.</span>
<span class="sd">        method (str): How to combine the replicate dataset.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of numeric dataset now merged together.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;Sum&quot;</span><span class="p">:</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;Mean&quot;</span><span class="p">:</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;TTRMean&quot;</span><span class="p">:</span>
        <span class="c1">#factors = norm_tools.TTR_factors(data)</span>
        <span class="c1">#data = factors * data</span>
        <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">factors</span><span class="p">)</span> <span class="o">=</span> <span class="n">norm_tools</span><span class="o">.</span><span class="n">normalize_data</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="s2">&quot;TTR&quot;</span><span class="p">)</span>
        <span class="n">target_factors</span> <span class="o">=</span> <span class="n">norm_tools</span><span class="o">.</span><span class="n">norm_to_target</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">target_factors</span> <span class="o">*</span> <span class="n">data</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">,:]</span>

    <span class="k">return</span> <span class="n">combined</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_data_stats"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_data_stats">[docs]</a><span class="k">def</span> <span class="nf">get_data_stats</span><span class="p">(</span><span class="n">reads</span><span class="p">):</span>
    <span class="n">density</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">reads</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">meanrd</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>
    <span class="n">nzmeanrd</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">reads</span><span class="p">[</span><span class="n">reads</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">nzmedianrd</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">reads</span><span class="p">[</span><span class="n">reads</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">maxrd</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>
    <span class="n">totalrd</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span>

    <span class="n">skew</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">skew</span><span class="p">(</span><span class="n">reads</span><span class="p">[</span><span class="n">reads</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">kurtosis</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">kurtosis</span><span class="p">(</span><span class="n">reads</span><span class="p">[</span><span class="n">reads</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">density</span><span class="p">,</span> <span class="n">meanrd</span><span class="p">,</span> <span class="n">nzmeanrd</span><span class="p">,</span> <span class="n">nzmedianrd</span><span class="p">,</span> <span class="n">maxrd</span><span class="p">,</span> <span class="n">totalrd</span><span class="p">,</span> <span class="n">skew</span><span class="p">,</span> <span class="n">kurtosis</span><span class="p">)</span></div>



<div class="viewcode-block" id="get_wig_stats"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_wig_stats">[docs]</a><span class="k">def</span> <span class="nf">get_wig_stats</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns statistics for the given wig file with read-counts.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): String with the path to the wig file of interest.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: Tuple with the following statistical measures:</span>
<span class="sd">            - density</span>
<span class="sd">            - mean read</span>
<span class="sd">            - non-zero mean</span>
<span class="sd">            - non-zero median</span>
<span class="sd">            - max read</span>
<span class="sd">            - total reads</span>
<span class="sd">            - skew</span>
<span class="sd">            - kurtosis</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">position</span><span class="p">)</span> <span class="o">=</span> <span class="n">get_data</span><span class="p">([</span><span class="n">path</span><span class="p">])</span>
    <span class="n">reads</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">get_data_stats</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_extended_pos_hash_pt"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_extended_pos_hash_pt">[docs]</a><span class="k">def</span> <span class="nf">get_extended_pos_hash_pt</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="c1">#TODO: Write docstring</span>

    <span class="nb">hash</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">maxcoord</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;-inf&quot;</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span> <span class="k">continue</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">orf</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">maxcoord</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">maxcoord</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
        <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">orf</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>

    <span class="n">genome_start</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">if</span> <span class="n">N</span><span class="p">:</span>
        <span class="n">genome_end</span> <span class="o">=</span> <span class="n">maxcoord</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">genome_end</span> <span class="o">=</span> <span class="n">N</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,(</span><span class="n">orf</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">genome_start</span> <span class="o">&gt;</span> <span class="n">start</span><span class="p">:</span>
            <span class="n">genome_start</span> <span class="o">=</span> <span class="n">start</span>

        <span class="n">prev_orf</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">prev_orf</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">next_orf</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">next_orf</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">genome_start</span><span class="p">,</span> <span class="n">end</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">hash</span><span class="p">:</span> <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;current&quot;</span><span class="p">:[],</span> <span class="s2">&quot;prev&quot;</span><span class="p">:[],</span> <span class="s2">&quot;next&quot;</span><span class="p">:[]}</span>

            <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;prev&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev_orf</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">pos</span> <span class="o">&gt;=</span> <span class="n">start</span><span class="p">:</span>
                <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;next&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">next_orf</span><span class="p">)</span>
                <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;current&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">orf</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;next&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">orf</span><span class="p">)</span>
        <span class="n">genome_start</span> <span class="o">=</span> <span class="n">end</span><span class="o">+</span><span class="mi">1</span>

    <span class="k">if</span> <span class="n">N</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">maxcoord</span><span class="p">,</span> <span class="n">genome_end</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">hash</span><span class="p">:</span> <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;current&quot;</span><span class="p">:[],</span> <span class="s2">&quot;prev&quot;</span><span class="p">:[],</span> <span class="s2">&quot;next&quot;</span><span class="p">:[]}</span>
            <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;prev&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev_orf</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">hash</span></div>

<div class="viewcode-block" id="get_extended_pos_hash_gff"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_extended_pos_hash_gff">[docs]</a><span class="k">def</span> <span class="nf">get_extended_pos_hash_gff</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="c1">#TODO: Write docstring</span>

    <span class="nb">hash</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">maxcoord</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;-inf&quot;</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span> <span class="k">continue</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">features</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">))</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)])</span>
        <span class="k">if</span> <span class="s2">&quot;ID&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span> <span class="k">continue</span>
        <span class="n">orf</span> <span class="o">=</span> <span class="n">features</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span>
        <span class="nb">chr</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
        <span class="n">maxcoord</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">maxcoord</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
        <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">orf</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">))</span>

    <span class="n">genome_start</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">if</span> <span class="n">N</span><span class="p">:</span>
        <span class="n">genome_end</span> <span class="o">=</span> <span class="n">maxcoord</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">genome_end</span> <span class="o">=</span> <span class="n">N</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,(</span><span class="n">orf</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">genome_start</span> <span class="o">&gt;</span> <span class="n">start</span><span class="p">:</span>
            <span class="n">genome_start</span> <span class="o">=</span> <span class="n">start</span>

        <span class="n">prev_orf</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">prev_orf</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">next_orf</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">next_orf</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">genome_start</span><span class="p">,</span> <span class="n">end</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">hash</span><span class="p">:</span> <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;current&quot;</span><span class="p">:[],</span> <span class="s2">&quot;prev&quot;</span><span class="p">:[],</span> <span class="s2">&quot;next&quot;</span><span class="p">:[]}</span>

            <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;prev&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev_orf</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">pos</span> <span class="o">&gt;=</span> <span class="n">start</span><span class="p">:</span>
                <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;next&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">next_orf</span><span class="p">)</span>
                <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;current&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">orf</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;next&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">orf</span><span class="p">)</span>
        <span class="n">genome_start</span> <span class="o">=</span> <span class="n">end</span><span class="o">+</span><span class="mi">1</span>

    <span class="k">if</span> <span class="n">N</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">maxcoord</span><span class="p">,</span> <span class="n">genome_end</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">hash</span><span class="p">:</span> <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;current&quot;</span><span class="p">:[],</span> <span class="s2">&quot;prev&quot;</span><span class="p">:[],</span> <span class="s2">&quot;next&quot;</span><span class="p">:[]}</span>
            <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="s2">&quot;prev&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prev_orf</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">hash</span></div>






<div class="viewcode-block" id="get_pos_hash_pt"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_pos_hash_pt">[docs]</a><span class="k">def</span> <span class="nf">get_pos_hash_pt</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a dictionary that maps coordinates to a list of genes that occur at that coordinate.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): Path to annotation in .prot_table format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary of position to list of genes that share that position.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">hash</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span> <span class="k">continue</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">orf</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">hash</span><span class="p">:</span> <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">orf</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">hash</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_pos_hash_gff"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_pos_hash_gff">[docs]</a><span class="k">def</span> <span class="nf">get_pos_hash_gff</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a dictionary that maps coordinates to a list of genes that occur at that coordinate.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): Path to annotation in GFF3 format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary of position to list of genes that share that position.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">hash</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span> <span class="k">continue</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">features</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;=&quot;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">,</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">))])</span>
        <span class="k">if</span> <span class="s2">&quot;ID&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span> <span class="k">continue</span>
        <span class="n">orf</span> <span class="o">=</span> <span class="n">features</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span>
        <span class="nb">chr</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">hash</span><span class="p">:</span> <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="nb">hash</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">orf</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">hash</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_pos_hash"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_pos_hash">[docs]</a><span class="k">def</span> <span class="nf">get_pos_hash</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a dictionary that maps coordinates to a list of genes that occur at that coordinate.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): Path to annotation in .prot_table or GFF3 format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary of position to list of genes that share that position.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filename</span><span class="p">,</span> <span class="n">file_extension</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">file_extension</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;.gff&quot;</span><span class="p">,</span> <span class="s2">&quot;.gff3&quot;</span><span class="p">]:</span>
        <span class="k">return</span> <span class="n">get_pos_hash_gff</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">get_pos_hash_pt</span><span class="p">(</span><span class="n">path</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_gene_info_pt"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_gene_info_pt">[docs]</a><span class="k">def</span> <span class="nf">get_gene_info_pt</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a dictionary that maps gene id to gene information.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): Path to annotation in .prot_table format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary of gene id to tuple of information:</span>
<span class="sd">            - name</span>
<span class="sd">            - description</span>
<span class="sd">            - start coordinate</span>
<span class="sd">            - end coordinate</span>
<span class="sd">            - strand</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">orf2info</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span> <span class="k">continue</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">orf</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
        <span class="n">desc</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
        <span class="n">orf2info</span><span class="p">[</span><span class="n">orf</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">desc</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">orf2info</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_gene_info_gff"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_gene_info_gff">[docs]</a><span class="k">def</span> <span class="nf">get_gene_info_gff</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a dictionary that maps gene id to gene information.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): Path to annotation in GFF3 format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary of gene id to tuple of information:</span>
<span class="sd">            - name</span>
<span class="sd">            - description</span>
<span class="sd">            - start coordinate</span>
<span class="sd">            - end coordinate</span>
<span class="sd">            - strand</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">orf2info</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span> <span class="k">continue</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">chr</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
        <span class="n">length</span> <span class="o">=</span> <span class="p">((</span><span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="mi">3</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
        <span class="n">features</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;=&quot;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">,</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">))])</span>
        <span class="k">if</span> <span class="s2">&quot;ID&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span> <span class="k">continue</span>
        <span class="n">orf</span> <span class="o">=</span> <span class="n">features</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="n">name</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>

        <span class="n">desc</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;Description&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">desc</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;description&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">desc</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;Desc&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">desc</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;desc&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">desc</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;product&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>

        <span class="n">orf2info</span><span class="p">[</span><span class="n">orf</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">desc</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">orf2info</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_gene_info"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_gene_info">[docs]</a><span class="k">def</span> <span class="nf">get_gene_info</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a dictionary that maps gene id to gene information.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): Path to annotation in .prot_table or GFF3 format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary of gene id to tuple of information:</span>
<span class="sd">            - name</span>
<span class="sd">            - description</span>
<span class="sd">            - start coordinate</span>
<span class="sd">            - end coordinate</span>
<span class="sd">            - strand</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filename</span><span class="p">,</span> <span class="n">file_extension</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">file_extension</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;.gff&quot;</span><span class="p">,</span> <span class="s2">&quot;.gff3&quot;</span><span class="p">]:</span>
        <span class="k">return</span> <span class="n">get_gene_info_gff</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">get_gene_info_pt</span><span class="p">(</span><span class="n">path</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_coordinate_map"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_coordinate_map">[docs]</a><span class="k">def</span> <span class="nf">get_coordinate_map</span><span class="p">(</span><span class="n">galign_path</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Attempts to get mapping of coordinates from galign file.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): Path to .galign file.</span>
<span class="sd">        reverse (bool): Boolean specifying whether to do A to B or B to A.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary of coordinate in one file to another file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">c1Toc2</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">galign_path</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span> <span class="k">continue</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="n">star</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span>
        <span class="n">leftempty</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="n">rightempty</span> <span class="o">=</span> <span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">leftempty</span><span class="p">:</span>
            <span class="n">left</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">left</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">rightempty</span><span class="p">:</span>
            <span class="n">right</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="k">elif</span> <span class="n">leftempty</span><span class="p">:</span>
            <span class="n">right</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
             <span class="n">right</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tmp</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">reverse</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">leftempty</span><span class="p">:</span>
                <span class="n">c1Toc2</span><span class="p">[</span><span class="n">left</span><span class="p">]</span> <span class="o">=</span> <span class="n">right</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">rightempty</span><span class="p">:</span>
                 <span class="n">c1Toc2</span><span class="p">[</span><span class="n">right</span><span class="p">]</span> <span class="o">=</span> <span class="n">left</span>
    <span class="k">return</span> <span class="n">c1Toc2</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="read_genome"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.read_genome">[docs]</a><span class="k">def</span> <span class="nf">read_genome</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads in FASTA formatted genome file.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        path (str): Path to .galign file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        string: String with the genomic sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">):</span> <span class="k">continue</span>
        <span class="n">seq</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">seq</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="maxrun"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.maxrun">[docs]</a><span class="k">def</span> <span class="nf">maxrun</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span><span class="n">item</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns the length of the maximum run an item in a given list.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        lst (list): List of numeric items.</span>
<span class="sd">        item (float): Number to look for consecutive runs of.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: Length of the maximum run of consecutive instances of item.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">best</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">i</span><span class="p">,</span><span class="n">n</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">lst</span><span class="p">)</span>
    <span class="k">while</span> <span class="n">i</span><span class="o">&lt;</span><span class="n">n</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">lst</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="n">item</span><span class="p">:</span>
            <span class="n">j</span> <span class="o">=</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span>
            <span class="k">while</span> <span class="n">j</span><span class="o">&lt;</span><span class="n">n</span> <span class="ow">and</span> <span class="n">lst</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">==</span><span class="n">item</span><span class="p">:</span> <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">r</span> <span class="o">=</span> <span class="n">j</span><span class="o">-</span><span class="n">i</span>
            <span class="k">if</span> <span class="n">r</span><span class="o">&gt;</span><span class="n">best</span><span class="p">:</span> <span class="n">best</span> <span class="o">=</span> <span class="n">r</span>
            <span class="n">i</span> <span class="o">=</span> <span class="n">j</span>
        <span class="k">else</span><span class="p">:</span> <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">best</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="getR1"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.getR1">[docs]</a><span class="k">def</span> <span class="nf">getR1</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Small Correction term. Defaults to 0.000016 for now&quot;&quot;&quot;</span>
    <span class="k">return</span><span class="p">(</span><span class="mf">0.000016</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="getR2"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.getR2">[docs]</a><span class="k">def</span> <span class="nf">getR2</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Small Correction term. Defaults to 0.00006 for now&quot;&quot;&quot;</span>
    <span class="k">return</span><span class="p">(</span><span class="mf">0.00006</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="getE1"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.getE1">[docs]</a><span class="k">def</span> <span class="nf">getE1</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Small Correction term. Defaults to 0.01 for now&quot;&quot;&quot;</span>
    <span class="k">return</span><span class="p">(</span><span class="mf">0.01</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="getE2"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.getE2">[docs]</a><span class="k">def</span> <span class="nf">getE2</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Small Correction term. Defaults to 0.01 for now&quot;&quot;&quot;</span>
    <span class="k">return</span><span class="p">(</span><span class="mf">0.01</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="getGamma"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.getGamma">[docs]</a><span class="k">def</span> <span class="nf">getGamma</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Euler-Mascheroni constant ~ 0.577215664901 &quot;&quot;&quot;</span>
    <span class="k">return</span><span class="p">(</span><span class="mf">0.5772156649015328606</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="ExpectedRuns"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.ExpectedRuns">[docs]</a><span class="k">def</span> <span class="nf">ExpectedRuns</span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="n">pnon</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Expected value of the run of non=insertions (Schilling, 1990):</span>

<span class="sd">        ER_n =  log(1/p)(nq) + gamma/ln(1/p) -1/2 + r1(n) + E1(n)</span>

<span class="sd">    Arguments:</span>
<span class="sd">        n (int): Integer representing the number of sites.</span>
<span class="sd">        pins (float): Floating point number representing the probability of non-insertion.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: Size of the expected maximum run.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">n</span><span class="o">&lt;</span><span class="mi">20</span><span class="p">:</span> <span class="c1"># use exact calculation for genes with less than 20 TA sites</span>
      <span class="c1"># Eqn 17-20 in Boyd, https://www.math.ubc.ca/~boyd/bern.runs/bernoulli.html</span>
      <span class="c1">#  recurrence relations for F(n,k) = prob that max run has length k</span>
      <span class="n">p</span><span class="p">,</span><span class="n">q</span> <span class="o">=</span> <span class="mi">1</span><span class="o">-</span><span class="n">pnon</span><span class="p">,</span><span class="n">pnon</span>
      <span class="n">F</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
      <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span> <span class="n">F</span><span class="p">[</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span><span class="o">-</span><span class="n">numpy</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="n">q</span><span class="p">,</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
      <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
          <span class="k">if</span> <span class="n">n</span><span class="o">&gt;=</span><span class="n">k</span><span class="o">+</span><span class="mi">2</span><span class="p">:</span> <span class="n">F</span><span class="p">[</span><span class="n">n</span><span class="p">,</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">F</span><span class="p">[</span><span class="n">n</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="n">k</span><span class="p">]</span><span class="o">-</span><span class="n">p</span><span class="o">*</span><span class="n">numpy</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="n">q</span><span class="p">,</span><span class="n">k</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">F</span><span class="p">[</span><span class="n">n</span><span class="o">-</span><span class="n">k</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="n">k</span><span class="p">]</span>
      <span class="n">ERn</span> <span class="o">=</span> <span class="mi">0</span>
      <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span> <span class="n">ERn</span> <span class="o">+=</span> <span class="n">k</span><span class="o">*</span><span class="p">(</span><span class="n">F</span><span class="p">[</span><span class="n">n</span><span class="p">,</span><span class="n">k</span><span class="p">]</span><span class="o">-</span><span class="n">F</span><span class="p">[</span><span class="n">n</span><span class="p">,</span><span class="n">k</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
      <span class="k">return</span> <span class="n">ERn</span>

    <span class="n">pins</span> <span class="o">=</span> <span class="mi">1</span><span class="o">-</span><span class="n">pnon</span>
    <span class="n">gamma</span> <span class="o">=</span> <span class="n">getGamma</span><span class="p">()</span>
    <span class="n">r1</span> <span class="o">=</span> <span class="n">getR1</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
    <span class="n">E1</span> <span class="o">=</span> <span class="n">getE1</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
    <span class="n">A</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">n</span><span class="o">*</span><span class="n">pins</span><span class="p">,</span><span class="mf">1.0</span><span class="o">/</span><span class="n">pnon</span><span class="p">)</span>
    <span class="n">B</span> <span class="o">=</span> <span class="n">gamma</span><span class="o">/</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">pnon</span><span class="p">)</span>
    <span class="n">ER</span> <span class="o">=</span> <span class="n">A</span> <span class="o">+</span> <span class="n">B</span> <span class="o">-</span><span class="mf">0.5</span> <span class="o">+</span> <span class="n">r1</span> <span class="o">+</span> <span class="n">E1</span>
    <span class="k">return</span> <span class="n">ER</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="VarR"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.VarR">[docs]</a><span class="k">def</span> <span class="nf">VarR</span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="n">pnon</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Variance of the expected run of non-insertons (Schilling, 1990):</span>

<span class="sd">    .. math::</span>

<span class="sd">        VarR_n =  (pi^2)/(6*ln(1/p)^2) + 1/12 + r2(n) + E2(n)</span>


<span class="sd">    Arguments:</span>
<span class="sd">        n (int): Integer representing the number of sites.</span>
<span class="sd">        pnon (float): Floating point number representing the probability of non-insertion.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: Variance of the length of the maximum run.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">r2</span> <span class="o">=</span> <span class="n">getR2</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
    <span class="n">E2</span> <span class="o">=</span> <span class="n">getE2</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
    <span class="n">A</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="p">,</span><span class="mf">2.0</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="mi">6</span><span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">pow</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">pnon</span><span class="p">),</span><span class="mf">2.0</span><span class="p">))</span>
    <span class="n">V</span> <span class="o">=</span> <span class="n">A</span> <span class="o">+</span> <span class="mi">1</span><span class="o">/</span><span class="mf">12.0</span> <span class="o">+</span> <span class="n">r2</span> <span class="o">+</span> <span class="n">E2</span>
    <span class="k">return</span> <span class="n">V</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="GumbelCDF"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.GumbelCDF">[docs]</a><span class="k">def</span> <span class="nf">GumbelCDF</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">u</span><span class="p">,</span><span class="n">B</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;CDF of the Gumbel distribution:</span>

<span class="sd">        e^(-e^( (u-x)/B))</span>

<span class="sd">    Arguments:</span>
<span class="sd">        x (int): Length of the max run.</span>
<span class="sd">        u (float): Location parameter of the Gumbel dist.</span>
<span class="sd">        B (float): Scale parameter of the Gumbel dist.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: Cumulative probability o the Gumbel distribution.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">((</span><span class="n">u</span><span class="o">-</span><span class="n">x</span><span class="p">)</span><span class="o">/</span><span class="n">B</span> <span class="p">)))</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="griffin_analysis"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.griffin_analysis">[docs]</a><span class="k">def</span> <span class="nf">griffin_analysis</span><span class="p">(</span><span class="n">genes_obj</span><span class="p">,</span> <span class="n">pins</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Implements the basic Gumbel analysis of runs of non-insertion, described in Griffin et al. 2011.</span>

<span class="sd">    This analysis method calculates a p-value of observing the maximun run of</span>
<span class="sd">    TA sites without insertions in a row (i.e. a &quot;run&quot;, r). Unusually long</span>
<span class="sd">    runs are indicative of an essential gene or protein domain. Assumes that</span>
<span class="sd">    there is a constant, global probability of observing an insertion</span>
<span class="sd">    (tantamount to a Bernoulli probability of success).</span>

<span class="sd">    Arguments:</span>
<span class="sd">        genes_obj (Genes): An object of the Genes class defining the genes.</span>
<span class="sd">        pins (float): The probability of insertion.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of lists with results and information for the genes. The elements of the list are as follows:</span>
<span class="sd">            - ORF ID.</span>
<span class="sd">            - Gene Name.</span>
<span class="sd">            - Gene Description.</span>
<span class="sd">            - Number of TA sites with insertions.</span>
<span class="sd">            - Number of TA sites.</span>
<span class="sd">            - Length of largest run of non-insertion.</span>
<span class="sd">            - Expected run for a gene this size.</span>
<span class="sd">            - p-value of the observed run.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">pnon</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">pins</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes_obj</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="o">.</span><span class="n">orf</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">desc</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">k</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">n</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">r</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.000</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># do I need to estimate B better (using exact calc for variance) for small genes too?</span>
            <span class="n">B</span> <span class="o">=</span> <span class="mf">1.0</span><span class="o">/</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">pnon</span><span class="p">)</span> <span class="c1"># beta param of Gumbel distn; like tau in our Bioinfo paper</span>
            <span class="c1">#u = math.log(gene.n*pins, 1.0/pnon) # instead, calculate this based on estimate of ExpectedRun() length below</span>
            <span class="n">exprun</span> <span class="o">=</span> <span class="n">ExpectedRuns</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">n</span><span class="p">,</span> <span class="n">pnon</span><span class="p">)</span>
            <span class="c1"># u is mu of Gumbel (mean=mu+gamma*beta); matching of moments; like Eq 5 in Schilling, but subtract off unneeded terms</span>
            <span class="n">u</span> <span class="o">=</span> <span class="n">exprun</span><span class="o">-</span><span class="n">getGamma</span><span class="p">()</span><span class="o">/</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">1.0</span><span class="o">/</span><span class="n">pnon</span><span class="p">)</span>
            <span class="n">pval</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">GumbelCDF</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">r</span><span class="p">,</span> <span class="n">u</span><span class="p">,</span> <span class="n">B</span><span class="p">)</span>
            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene</span><span class="o">.</span><span class="n">orf</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">desc</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">k</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">n</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">r</span><span class="p">,</span> <span class="n">exprun</span><span class="p">,</span> <span class="n">pval</span><span class="p">])</span>
    <span class="k">return</span><span class="p">(</span><span class="n">results</span><span class="p">)</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="runs_w_info"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.runs_w_info">[docs]</a><span class="k">def</span> <span class="nf">runs_w_info</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return list of all the runs of consecutive non-insertions with the start and end locations.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        data (list): List of numeric data to check for runs.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of dictionary from run to length and position information of the tun.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">runs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">start</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">current_r</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">read</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># If ending a run of zeros</span>
            <span class="k">if</span> <span class="n">current_r</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># If we were in a run, add to list</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">current_r</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="n">runs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">current_r</span><span class="p">,</span> <span class="n">start</span> <span class="o">=</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">end</span><span class="p">))</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="p">(</span><span class="n">current_r</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">current_r</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">current_r</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="c1"># If we ended in a run, add it</span>
    <span class="k">if</span> <span class="n">current_r</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">current_r</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">runs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">current_r</span><span class="p">,</span> <span class="n">start</span> <span class="o">=</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">end</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">runs</span></div>

<span class="c1">#</span>

<div class="viewcode-block" id="get_genes_in_range"><a class="viewcode-back" href="../../transit.html#pytransit.tnseq_tools.get_genes_in_range">[docs]</a><span class="k">def</span> <span class="nf">get_genes_in_range</span><span class="p">(</span><span class="n">pos_hash</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns list of genes that occur in a given range of coordinates.</span>

<span class="sd">    Arguments:</span>
<span class="sd">        pos_hash (dict): Dictionary of position to list of genes.</span>
<span class="sd">        start (int): Start coordinate of the desired range.</span>
<span class="sd">        end (int): End coordinate of the desired range.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of genes that fall within range.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">pos_hash</span><span class="p">:</span>
            <span class="n">genes</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">pos_hash</span><span class="p">[</span><span class="n">pos</span><span class="p">])</span>

    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">genes</span><span class="p">))</span></div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>

    <span class="n">G</span> <span class="o">=</span> <span class="n">Genes</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">),</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">norm</span><span class="o">=</span><span class="s2">&quot;TTR&quot;</span><span class="p">)</span>
    <span class="n">theta</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">global_theta</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;#Insertion: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">G</span><span class="o">.</span><span class="n">global_insertion</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;#Sites: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">G</span><span class="o">.</span><span class="n">global_sites</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;#Run: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">G</span><span class="o">.</span><span class="n">global_run</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;#Theta: </span><span class="si">%1.4f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">theta</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;#Phi: </span><span class="si">%1.4f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">G</span><span class="o">.</span><span class="n">global_phi</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">)</span>

    <span class="n">griffin_results</span> <span class="o">=</span> <span class="n">griffin_analysis</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">theta</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">position</span>
        <span class="n">exprun</span><span class="p">,</span> <span class="n">pval</span> <span class="o">=</span> <span class="n">griffin_results</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="o">-</span><span class="mi">2</span><span class="p">:]</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%1.1f</span><span class="se">\t</span><span class="si">%1.5f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">orf</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">k</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">n</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">r</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">t</span><span class="p">,</span> <span class="n">exprun</span><span class="p">,</span> <span class="n">pval</span><span class="p">))</span>


</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2015, Michael A. DeJesus &amp; Thomas R. Ioerger.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>